### ISSUE TITLE
GPG encryption failing with No public key Error

### ISSUE THREAD

**Issue by SirishaGopigiri** [mozilla/sops#767]
Hi

We are trying to use sops with gpg to encrypt some of the yaml files. We are using the sops golibrary to perform the encryption and decryption functionality on the YAML files. We have built a sample public and private keys and have imported the private key using `gpg --import ` command. The keys are imported successfully and are available through `gpg --list-keys` command.

```
------------------------
pub   rsa2048 2020-08-21 [SC] [expires: 2022-08-21]
      681E3A89EB1DAFD36EB883120A73BB48E26694D8
uid           [ unknown] Docker <Docker@cluster.local>
sub   rsa2048 2020-08-21 [E] [expires: 2022-08-21]
```

But when we are trying to use the same key to encrypt a file, it seems to be failing with the following error:

```
failed to encrypt new data key with master key "681E3A89EB1DAFD36EB883120A73BB48E26694D8": could not encrypt data key with PGP key: golang.org/x/crypto/openpgp error: key with fingerprint 681E3A89EB1DAFD36EB883120A73BB48E26694D8 is not available in keyring and could not be retrieved from keyserver; GPG binary error: gpg binary failed with error: exit status 2, gpg: 681E3A89EB1DAFD36EB883120A73BB48E26694D8: skipped: No public key
gpg: [stdin]: encryption failed: No public key
```

We are expecting that the execution is failing at this step [](https://github.com/mozilla/sops/blob/develop/pgp/keysource.go#L83)

The same keys seems to be working fine most of the time, it is failing with the above error 5 out of 10 times when executed. As the error is not consistent we are not sure what are we missing here.

can someone help us identify what exactly is the issue here?

Thank you in advance!

**Comment by felixfontein** [comment]
This looks like gpg sometimes fails. I sometimes have a maybe related problem, which happens if I simulatenously (in different forks) call sops to decrypt a (or multiple) files. Then sometimes some of the calls end up with something like:
```
Failed to get the data key required to decrypt the SOPS file.

Group 0: FAILED
  D13xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx: FAILED
    - | could not decrypt data key with PGP key:
      | golang.org/x/crypto/openpgp error: Reading PGP message
      | failed: openpgp: incorrect key; GPG binary error: exit
      | status 2
  
  828xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx: FAILED
    - | could not decrypt data key with PGP key:
      | golang.org/x/crypto/openpgp error: Reading PGP message
      | failed: openpgp: incorrect key; GPG binary error: exit
      | status 2

Recovery failed because no master key was able to decrypt the file. In
order for SOPS to recover the file, at least one key has to be successful,
but none were.
```
Maybe gpg - or gpg-agent - isn't happy when it is called in parallel?

**Comment by SirishaGopigiri** [comment]
Thank you for the quick reply @felixfontein 

But I don't think they are happening in parallel. The execution is failing as part of testcases execution for the gocode written using sops library. Some testcases are failing with the above error message while some are passing.

**Comment by felixfontein** [comment]
Do you run the go tests in parallel? It seems like you can explicitly disable it with `go test -parallel=1`.

Anyway, I was able to trigger errors during gpg decryption when running it in semi-parallel:
```
for i in 1 2 3 4 5 6 7 8 9 10; do ( gpg --use-agent --decrypt test.gpg > /dev/null & ); done
```
In almost every run, at least one of the calls ends with
```
gpg: public key decryption failed: Cannot allocate memory
gpg: decryption failed: No secret key
```
If I leave `--use-agent` away, it works flawlessy. (sops always specifies `--use-agent -d`: https://github.com/mozilla/sops/blob/master/pgp/keysource.go#L186-L201)

**Comment by felixfontein** [comment]
Now I found https://dev.gnupg.org/T4255, and adding `auto-expand-secmem` to `~/.gnupg/gpg-agent.conf` solved my problem for me. Would be interesting to know whether this (or just running tests in serial) helps in your case.

**Comment by SirishaGopigiri** [comment]
Thank you for the pointers @felixfontein will try it out.

But in our case we are getting the error while encryption as well.

**Comment by felixfontein** [comment]
If I understand this problem correctly it can happen both for encryption and decryption.

**Comment by SirishaGopigiri** [comment]
Hi @felixfontein sorry for the late reply yes it happens for both encryption and decryption.

And I have tested both the solutions as suggested

1. using '-parallel=1' flag
2. adding auto-expand-secmem to ~/.gnupg/gpg-agent.conf

Unfortunately both of them haven't worked.

**Comment by felixfontein** [comment]
@SirishaGopigiri thanks for checking! I was suspecting a problem with gpg-agent since this error happens for you only sometimes. So I guess it's either a different gpg-agent problem, or something completely different. If it would always happen in the same situations, it would probably be a lot easier to analyze...

**Comment by SirishaGopigiri** [comment]
@felixfontein Yes agreed we are having trouble to debug it as it is not frequent but not sure why is it failing in some cases only

**Comment by dbartholomae** [comment]
Had a similar issue, and using a key without a passphrase solved it.
